### ISSUE TITLE
Conflicts with Magento Base API 

### ISSUE THREAD

**Issue by Feyks** [lizardmedia/product-attachments-magento2#9]
Hi,

I have spotted one bugs linked to the Magento API and blocking updates of products.

## Context

- Magento 2.3.3 - REST API

## Step to reproduce

1. Launch the following request : 
- PUT \<magento-url>/rest/V1/products/\<product-sku>
- `{
  "product": {
    "price": 5.27
  }
}`
## Expecting Result
- Product Data

### Having

> Fatal Error: 'Uncaught TypeError: Return value of LizardMedia\ProductAttachment\Model\AttachmentRepository::getAttachments() must be an instance of LizardMedia\ProductAttachment\Model\ResourceModel\Attachment\Collection or null, array returned in /home/magento/fm-medical/app/code/LizardMedia/ProductAttachment/Model/AttachmentRepository.php:255

**Comment by Feyks** [comment]
The second bug was fixed in the last version, i just installed an old version. But bug one persists.

**Comment by bartoszkubicki** [comment]
@Feyks Thank you for your report. I hope it will be fixed in the begining of january. Of course you can post PR, I bet it is not very complicated issue.

**Comment by Feyks** [comment]
@bartoszkubicki I'm still trying to fix this issue but I'm getting nowhere. I had a lead with type declaration after function names but it created more errors afterwards.

**Comment by bartoszkubicki** [comment]
@Feyks I will fix it as soon as possible

**Comment by bartoszkubicki** [comment]
@Feyks I am sorry but I am not able to reproduce this issue. I have tried for products having attachments and for products not having attachements. It doesnt't seem possible to me that this function returns an array. Even if `$product` doesn't have any attachement, empty collection should be attached. Code is not very new and requires refactor and simplifying, but this part seem to work. 

Make sure that your code is actual, and repoen this issue if you have any more guidance to reproduce it.

**Comment by Feyks** [comment]
I have some news on this issue. Trying to reinstall by composer the module, i saw that the PHP requirements was incompatible. I was sure i had PHP v7.1 but it was upgraded to 7.3.11.
So I think the problem is coming from the upgrade of my PHP version.

**Comment by bartoszkubicki** [comment]
@Feyks module is going to be refactored in february a little bit and then we will add support for `7.2`. For now you can try installing it with `--ignore-platform-reqs`, but I cannot assure you that it is goint to work with 7.3.

**Comment by Feyks** [comment]
On 7.3 it won't work. I tried to download the ZIP file but had troubles. I m trying to install it with the ignore platform reqs option on the 7.1 to avoid symfony requirements errors.

**Comment by Feyks** [comment]
I succeded to make the API working with the following : 

### Replace

> Lizardmedia\ProductAttachments\Model\AttachmentRepository.php

```
/**
     * @param ProductInterface $product
     * @return AttachmentCollection
     * @throws Exception
     */
    private function getAttachments(ProductInterface $product) : ?Collection
    {
        if ($product->getProductAttachments() === null) {
            $attachmentCollection = $this->attachmentCollectionFactory->create()
                ->addProductToFilter([$product->getEntityId()])
                ->addTitleToResult((int) $product->getStoreId());
            $this->extensionAttributesJoinProcessor->process($attachmentCollection);
            $product->setProductAttachments($attachmentCollection);
        }
        return $product->getProductAttachments();
    }
```
### By

> Lizardmedia\ProductAttachments\Model\AttachmentRepository.php
```
/**
     * @param ProductInterface $product
     * @return AttachmentCollection
     * @throws Exception
     */
    private function getAttachments(ProductInterface $product) : ?Collection
    {
        if ($product->getProductAttachments() === null) {
            $attachmentCollection = $this->attachmentCollectionFactory->create()
                ->addProductToFilter([$product->getEntityId()])
                ->addTitleToResult((int) $product->getStoreId());
            $this->extensionAttributesJoinProcessor->process($attachmentCollection);
            $product->setProductAttachments($attachmentCollection);
            return $attachmentCollection;
        }
        return $product->getProductAttachments();
    }
```

I simply don't know why this modification made this working. I hope you will find a better solution.
Thanks for your time,

**Comment by bartoszkubicki** [comment]
@Feyks Hmm seems strange to me. If we make new `$collection` object by factory and set it on `$product` with use of magic funciton `__call`, which is in fact `setProductAttachements` function there always should be `$collection`. Actually nullable `Collection` return type is too broad, as we should always have collection on product, even if it's empty. I don't know why it's not working for you. Maybe I will go back to this place with scheduled refactor, or completly rebuild this, so maybe this will solve any problems.

**Comment by Feyks** [comment]
I tried to show the class of returned element in getProductAttachment with get_class. It showed the expected class Collection and is_array returned false. I really don't understand what's going on. I'll persist to look for the source of the problem and find a solution.

**Comment by Feyks** [comment]
I have continued to work and debug this issue and i know now what was going on : 

While saving product Attachments, the data saved was an array instead of Collection.
So at line 236 of Lizardmedia\ProductAttachments\Model\AttachmentRepository.php, Magento fetched an array and continued to work with it.

The only way i have found to make this right is modifying the following line to regenerate data if it had not been saved properly:

> Lizardmedia\ProductAttachments\Model\AttachmentRepository.php
```
    /**
     * @param ProductInterface $product
     * @return AttachmentCollection
     * @throws Exception
     */
    private function getAttachments(ProductInterface $product) : ?Collection
    {
        if ($product->getProductAttachments() === null || !is_a($product->getProductAttachments(),'LizardMedia\ProductAttachment\Model\ResourceModel\Attachment\Collection')) {
            $attachmentCollection = $this->attachmentCollectionFactory->create()
                ->addProductToFilter([$product->getEntityId()])
                ->addTitleToResult((int) $product->getStoreId());

            $this->extensionAttributesJoinProcessor->process($attachmentCollection);
            $product->setProductAttachments($attachmentCollection);
        }
        return $product->getProductAttachments();
```

**Comment by bartoszkubicki** [comment]
@Feyks but is this situation while using actual codebase? I have tested api call against product with and without attachment. Is this happening if you create product and then try to update it somehow using api?

**Comment by Feyks** [comment]
I don't know how this happened but correcting it had worked.

I just have one more question, can you make an API call without product_attachment ? Because without the attachment part of the product, Magento is trying to create a new row in the database for the title with exactly the same information as before.

**Comment by bartoszkubicki** [comment]
@Feyks Thats good to hear. 

I am not sure what situation are your trying to describe. You make API call without any info about attachement and what is the problem? Any exception or just this attribute for the title is saved even it it is not necessary?

**Comment by Feyks** [comment]
Sorry for being late, my mission ended in January. So what was the problem ?
First i added a file through the back office. It was working fine. 

Next, I updated a product info (price, name, ...) through the API. It worked fine too.

Then I updated once again through the API and the product became inaccessible. Orders that contained concerned products broke.

In order to discover what was happening I took a look in the database and found the problem. There was several entries in the product_assignement_title table for one store for concerned products. All broken products had several entries for targeted store.

So I invastigated the reason of this multiple entries. I found two cases : 
-> If the product_assignment is not given in the API : Creation of a new entry in title table (if there was already existing one, the product ceased to work)
-> If the product_assignment is given in the API : Expected behaviou happened, the product is updated and the assignment don't cause any trouble.
